<launch>
    <arg name="use_cuda" default="true" />
    <arg name="show" default="false" />
    <arg name="mapping_type" default="tsdf" />
    <arg name="texture_enabled" default="false" />
    <arg name="enable_mesher" default="false" />
    <arg name="output_map" default="false" />
    <arg name="max_disp_particles" default="10485760" />
    <arg name="max_mesh" default="3000000" />
    <arg name="max_ray_length" default="3.1" />
    <arg name="min_ray_length" default="0.3" />
    <arg name="vins" default="false" />
    <arg name="viz" default="false" />
    <arg name="enable_submap" default="true" />
    <arg name="drone_id" default="0" />
    <arg name="vins_config" default="/home/xuhao/bags/swarm_raw_parallel2_2021-10-18_17-09/Configs/SwarmConfig1/fisheye_ptgrey_n3/fisheye_cuda.yaml" />
    <arg name="skeleton_graph_gen" default="false" />
    
    <node pkg="taichislam" type="taichislam_node.py" name="taichislam_node_$(arg drone_id)" output="screen" >
        <param name="use_cuda" value="$(arg use_cuda)" type="boolean" />
        <param name="enable_rendering" value="$(arg show)" type="boolean" />
        <param name="mapping_type" value="$(arg mapping_type)" type="string" />
        <param name="texture_enabled" value="$(arg texture_enabled)" type="boolean" />
        <param name="enable_mesher" value="$(arg enable_mesher)" type="boolean" />
        <param name="enable_submap" value="$(arg enable_submap)" type="boolean" />
        <param name="output_map" value="$(arg output_map)" type="boolean" />
        <param name="drone_id" value="$(arg drone_id)" type="int" />
        <param name="enable_skeleton_graph_gen" value="$(arg skeleton_graph_gen)" type="boolean" />
        
        <param name="max_ray_length" value="$(arg max_ray_length)" type="double" />
        <param name="min_ray_length" value="$(arg min_ray_length)" type="double" />
    
        <param name="disp/max_disp_particles" value="$(arg max_disp_particles)" type="int" />
        <param name="disp/max_mesh" value="$(arg max_mesh)" type="int" />

        <remap from="~depth" to="/camera/depth/image_rect_raw" />
        <!-- <remap from="~pose" to="/vins_estimator/camera_pose" /> -->
        <remap from="~pose" to="/d2vins/camera_pose_1" />
        <remap from="~traj" to="/d2pgo/pgo_traj" />
        <remap from="~frame_local" to="/d2vins/frame_local" />
        <remap from="~image" to="/camera/infra1/image_rect_raw/compressed" />
        <!-- <remap from="~image" to="/camera/color/image_raw/compressed" /> -->

        <rosparam>
            disp_ceiling: 1.8
            disp_floor: -0.5
            texture_compressed: true
            voxel_scale: 0.1
            color_same_proj: true
            disp:
                res_x: 1920
                res_y: 1080
            Kdepth:
                fx: 384.2377014160156
                fy: 384.2377014160156
                cx: 323.4873046875
                cy: 235.0628204345703
            skeleton_graph_gen:
                max_raycast_dist: 3.0
                coll_det_num: 64
                frontier_combine_angle_threshold: 20
            Kcolor:
                fx: 604.7939453125
                fy: 604.9515991210938
                cx: 321.3017578125
                cy: 242.9977264404297
        </rosparam>
    </node>
    
    <include file="$(find vins)/launch/fisheye_node.launch" if="$(arg vins)">
        <arg config_file="$(arg vins_config)" />
    </include>

    <node name="rvizvisualisation" pkg="rviz" type="rviz" output="log" 
        args="-d $(find TaichiSLAM)/config/taichi-slam.rviz" 
        if="$(arg viz)"/>

</launch>